core/src/main/java/io/substrait/expression/WindowBound.java

```diff
@@ -5,58 +5,62 @@
 @Value.Enclosing
 public interface WindowBound {
```

Comment:
Rather than encoding the WindowBound type using a combination of BoundedKind and Direction, I've switched this to 4 classes:
1. Preceding
2. Following
3. Current
4. Unbounded

and introduced a WindowBoundVisitor to handle dispatch when converting from these to various other representations.

This representation of Window Bounds:
1. More closely matches the spec.
2. Reduces the need for casting of WindowBound interfaces to concrete types.
3. Reduces cognitive complexity (in my opinion) as users no longer need to reason about 6 different enum combinations (3 bounded kinds * 2 directions), and instead have 4 concrete classes.